src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -113,6 +115,245 @@ void uct_rdmacm_cm_cqs_cleanup(uct_rdmacm_cm_t *cm)
+                      UCT_IB_MLX5_HCA_CAP_OPMOD_GET_CUR);
+    ret = mlx5dv_devx_general_cmd(verbs, in, sizeof(in),
+                                  out, sizeof(out));
+    if (ret != 0) {
+        ctx->support = false;
+        ucs_debug("mlx5dv_devx_general_cmd(QUERY_HCA_CAP) failed: %m");
+        return UCS_ERR_UNSUPPORTED;
+    }
+    ctx->general_obj_types_caps = UCT_IB_MLX5DV_GET64(cmd_hca_cap, cap,
+                                                      general_obj_types);
+    if (!(ctx->general_obj_types_caps &
+        UCS_BIT(UCT_IB_MLX5_OBJ_TYPE_RESERVED_QPN))) {
+        ucs_debug("general_obj_types_caps: reserved qpn is not support");
+        ctx->support = false;
+        return UCS_ERR_UNSUPPORTED;
+    }
+    ctx->support = true;
+    return UCS_OK;
+#else
+    ctx->support = false;
+    return UCS_ERR_UNSUPPORTED;
+#endif
+}
+
+static ucs_status_t
+uct_rdmacm_cm_query_reserved_qpn_cap_2(struct ibv_context *verbs,
+                                       struct uct_rdmacm_cm_reserved_qpn_context *ctx)
+{
+#if HAVE_DECL_MLX5DV_IS_SUPPORTED
+    char out[UCT_IB_MLX5DV_ST_SZ_BYTES(query_hca_cap_out)] = {};
+    char in[UCT_IB_MLX5DV_ST_SZ_BYTES(query_hca_cap_in)]   = {};
+    void *cap;
+    int ret;
+
+    if (!mlx5dv_is_supported(verbs->device)) {
+        ctx->support = false;
+        ucs_debug("mlx5dv is not supported");
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    cap = UCT_IB_MLX5DV_ADDR_OF(query_hca_cap_out, out, capability);
+    UCT_IB_MLX5DV_SET(query_hca_cap_in, in, opcode,
+                      UCT_IB_MLX5_CMD_OP_QUERY_HCA_CAP);
+    UCT_IB_MLX5DV_SET(query_hca_cap_in, in, op_mod,
+                      (UCT_IB_MLX5_CAP_2_GENERAL << 1) |
+                      UCT_IB_MLX5_HCA_CAP_OPMOD_GET_CUR);
+    ret = mlx5dv_devx_general_cmd(verbs, in, sizeof(in),
+                                  out, sizeof(out));
+    if (ret != 0) {
+        ctx->support = false;
+        ucs_debug("mlx5dv_devx_general_cmd(QUERY_HCA_CAP_2) failed: %m");
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    ctx->support = true;
+    ctx->log_reserved_qpn_granularity = UCT_IB_MLX5DV_GET(cmd_hca_cap_2, cap,
+                                                          log_reserved_qpn_granularity);
+    ucs_debug("reserved qpn cap: log_reserved_qpn_granularity is 0x%x",
+              ctx->log_reserved_qpn_granularity);
+    return UCS_OK;
+#else
+    ctx->support = false;
+    return UCS_ERR_UNSUPPORTED;
+#endif
+}
+
+static ucs_status_t
+uct_rdmacm_cm_reserved_qpn_blk_list_free(struct uct_rdmacm_cm_reserved_qpn_context *ctx)
+{
+    ucs_status_t status = UCS_OK;
+
+#if HAVE_DECL_MLX5DV_IS_SUPPORTED
+    struct uct_rdmacm_cm_reserved_qpn_blk *blk, *tmp;
+
+    pthread_mutex_lock(&ctx->mutex);
+
+    ucs_list_for_each_safe(blk, tmp, &ctx->blk_list, entry) {
+        ucs_list_del(&blk->entry);
+        status = uct_rdmacm_cm_reserved_qpn_blk_free(blk);
+        if (status != UCS_OK) {
+            goto end;
```

Comment:
we use `out` label name in similar code, in case of `ucs_list_for_each_safe` could just `break;`, but actually I think we should print error/warn/diag and continue to avoid memory leak as you do in `uct_rdmacm_cm_reserved_qpn_context_cleanup`

Reply:
@evgeny-leksikov Fixed. Thanks!